overlord/devicestate/devicemgr.go

```diff
@@ -1385,3 +1391,66 @@ func checkFDEFeatures(st *state.State, kernelInfo *snap.Info) error {
 	//       returns {"error":"hardware-unsupported"} we don't.
 	return nil
 }
+
+func (m *DeviceManager) runFDESetupHook(op string, params *boot.FdeSetupHookParams) ([]byte, error) {
+	hooksup := &hookstate.HookSetup{
+		Snap:     params.KernelInfo.InstanceName(),
+		Revision: params.KernelInfo.Revision,
+		Hook:     "fde-setup",
+		// XXX: should this be configurable somehow?
+		Timeout: 5 * time.Minute,
+	}
+	req := &hookstate.FDESetupRequest{
+		Op:      op,
+		Key:     params.Key,
+		KeyName: params.KeyName,
+		Model: map[string]string{
+			"series":     params.Model.Series(),
+			"brand-id":   params.Model.BrandID(),
+			"model":      params.Model.Model(),
+			"grade":      string(params.Model.Grade()),
+			"signkey-id": params.Model.SignKeyID(),
+		},
+		// TODO: include boot chains
+
+	}
+	contextData := map[string]interface{}{
+		"fde-setup-request": req,
+	}
+	context, err := m.hookMgr.EphemeralRunHook(context.Background(), hooksup, contextData)
+	if err != nil {
+		return nil, fmt.Errorf("cannot run fde-setup hook for %s: %v", op, err)
+	}
+
+	// the hook is expected to call "snapctl fde-setup-result" which
+	// wil set the "fde-setup-result" value on the task
+	var hookResult []byte
+	context.Lock()
+	err = context.Get("fde-setup-result", &hookResult)
+	context.Unlock()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get sealed key result in fde-setup-hook %s: %v", op, err)
```

Comment:
this sounds too specific given that the result could be from features etc ?

Reply:
Indeed, fixed now.